packages/shared/src/server/repositories/events.ts

```diff
@@ -295,34 +601,52 @@ const getObservationsFromEventsTableInternal = async <T>(
     },
   ]);
 
-  const query = `
-      ${scoresCte},
-      ${tracesCte}
-      SELECT
-       ${selectString}
-      FROM events e
-        ${
-          traceTableFilter.length > 0 || orderByTraces || search.query
-            ? "LEFT JOIN traces t ON t.id = e.trace_id AND t.project_id = e.project_id"
-            : ""
-        }
-        ${hasScoresFilter ? `LEFT JOIN scores_agg AS s ON s.trace_id = e.trace_id and s.observation_id = e.span_id` : ""}
-      WHERE ${appliedObservationsFilter.query}
-        ${search.query}
-      ${chOrderBy}
-      ${limit !== undefined && offset !== undefined ? `LIMIT ${limit} OFFSET ${offset}` : "LIMIT 1000"};`;
+  // Build query using EventsQueryBuilder
+  const queryBuilder = new EventsQueryBuilder();
+
+  if (opts.select === "count") {
+    queryBuilder.selectFieldSet("count");
+  } else {
+    queryBuilder.selectFieldSet("base", "calculated");
+    if (selectIOAndMetadata) {
+      queryBuilder.selectFieldSet("io", "metadata");
+    }
+  }
+
+  queryBuilder
+    .when(hasScoresFilter, (b) =>
+      b.withCTE("scores_agg", eventsScoresAggregation, {
+        projectId,
+        startTimeFrom,
+      }),
+    )
+    .when(!!needsTraceJoin, (b) =>
```

Comment:
Functionally equivalent, but I often like this better for readability. Up to you and your preferences.
```suggestion
    .when(Boolean(needsTraceJoin), (b) =>
```

Reply:
I don't have a preference yet.